### ISSUE TITLE
Host Process crash with `TransactionContext Error: Cannot create index with outstanding updates`

### ISSUE THREAD

**Issue by aarashy** [duckdb/duckdb#3666]
#### What happens?
I'm building a server that manages a collection of different DuckDB files. Upon server startup, it: 
1. Opens connections to all of them. 
2. If certain tables are not already created in these DBs, run some `create table` statements that have primary keys (running an initial migration).

I recently had an error where server startup was failing due to the following error which was killing the server process:
```
terminate called after throwing an instance of 'duckdb::TransactionException'
   what(): TransactionContext Error: Cannot create index with outstanding updates
```
My `create table` migrations are wrapped in a transaction, but there is not a single "update" statement in any of these transactions. 

I'm specifically wondering why my host process would crash here, I can tolerate errors that do not crash my process.

#### To Reproduce
Unfortunately, this was my first time seeing this error and I lost the relevant database file. It seems like the database must've gotten into a bad state, but I don't know exactly how and I don't know how to reproduce it.

Obviously, running the following set of commands also triggers the error message, but it doesn't seem to crash my DuckDB CLI in the same way that my process was crashing:
```
D begin transaction;
D update foo set c1=4;
D create index foo_idx on foo (c1);
Error: TransactionContext Error: Cannot create index with outstanding updates
```

I wonder if there was some crazy race where my server / host process died in the middle of an update transaction and the local update state became corrupted. After which, whenever I tried to open the file during server startup, the server would throw the above error.

I know this information doesn't give you much to go on, so it's understandable if we have to close this without making much progress. Again, my highest priority is understanding why it could have caused my process to crash.

#### Environment (please complete the following information):
 - OS: Monterey 12.2, but I think this was running in my ubuntu docker container.
 - DuckDB Version: It was DuckDB version 0.2.9 which hit this bug.
 - DuckDB Client: Rust

**Comment by hannes** [comment]
My guess would be the rust driver not catching the exception on startup. It should.

**Comment by Mytherin** [comment]
Thanks for the report!

Unfortunately this issue is not particularly actionable - especially as it is (1) running on a very old version, and (2) using an unofficial client. As Hannes mentioned the crash is not actually caused by the system (as an exception is thrown) but rather caused by the client not correctly catching and handling the exception.

Feel free to re-open the issue if you encounter this issue again or if you have any way of reproducing the issue for us.

**Comment by aarashy** [comment]
Hi @hannes and @Mytherin --

I found a way to reproduce this panic on v0.4.0! I have a database file + WAL which, whenever you try to open it using the DuckDB CLI, the process crashes. 

```
D .open /corrupted_database_file
libc++abi: terminating with uncaught exception of type duckdb::TransactionException: TransactionContext Error: Cannot create index with outstanding updates
Abort trap: 6
logout
```

As I originally suspected, this indicates that the crash is really coming from DuckDB, not the Rust client! 

I'm 100% sure that this is related to the WAL, checkpointing, and having lots of Add Column/Drop Column for the following reasons:

1. I observed that my WAL file was bigger than my database file
db file size: 12288 bytes
wal file size: 827033 bytes

2. The way that I produced this corrupted file was by submitting many SQL statements computationally. 
About 1800 transactions are opened and then committed without any manual checkpointing. When I added manual checkpointing between every transaction, the same sequence of SQL statements didn't trigger this corrupted state!

3. The sequence of SQL statements which seemed to be submitted last before triggering this corrupted state were a series of `Drop Column + Add Column` in one transaction, immediately followed by row mutations on the same table in the following transaction. Could it be that the `Cannot create index` is related to the creation of ARTs?

I want to send my corrupted database file to you to help you debug it, but I would prefer not to post it publicly here. How should I send my file to you?
